diff --git a/app/controllers/assessments_controller.rb b/app/controllers/assessments_controller.rb
index cb17ce1b..ddfd8fa0 100755
--- a/app/controllers/assessments_controller.rb
+++ b/app/controllers/assessments_controller.rb
@@ -187,22 +187,27 @@ class AssessmentsController < ApplicationController
     # If all requirements are satisfied, extract assessment files.
     begin
       course_root = Rails.root.join("courses", @course.name)
+      assessment_path = Rails.root.join("courses/#{@course.name}/#{@assessment.name}")
       tar_extract.rewind
       tar_extract.each do |entry|
         relative_pathname = entry.full_name
+        entry_file = File.join(course_root, relative_pathname)
+        # Ensure file will lie within course, otherwise skip
+        next unless Archive.in_dir?(Pathname(entry_file), Pathname(assessment_path))
+
         if entry.directory?
-          FileUtils.mkdir_p(File.join(course_root, relative_pathname),
+          FileUtils.mkdir_p(entry_file,
                             mode: entry.header.mode, verbose: false)
         elsif entry.file?
           FileUtils.mkdir_p(File.join(course_root, File.dirname(relative_pathname)),
                             mode: entry.header.mode, verbose: false)
-          File.open(File.join(course_root, relative_pathname), "wb") do |f|
+          File.open(entry_file, "wb") do |f|
             f.write entry.read
           end
-          FileUtils.chmod entry.header.mode, File.join(course_root, relative_pathname),
+          FileUtils.chmod entry.header.mode, entry_file,
                           verbose: false
         elsif entry.header.typeflag == "2"
-          File.symlink entry.header.linkname, File.join(course_root, relative_pathname)
+          File.symlink entry.header.linkname, entry_file
         end
       end
       tar_extract.close
diff --git a/app/controllers/courses_controller.rb b/app/controllers/courses_controller.rb
index d4756eae..451ddde9 100755
--- a/app/controllers/courses_controller.rb
+++ b/app/controllers/courses_controller.rb
@@ -1110,9 +1110,9 @@ private
 
         output_dir = if archive
                        extFilesDir
-                      else
-                        baseFilesDir
-                      end
+                     else
+                       baseFilesDir
+                     end
         output_file = File.join(output_dir, pathname)
 
         # skip if the file lies outside the archive
